src/vos/tests/csum_extent_tests.c

```diff
@@ -942,3 +942,80 @@ void csum_invalid_input_tests(void **state)
 	free(read_data_buf);
 	free(csum_buf_1);
 }
+
+/* Inject fault on the update and fetch.
+* The update will initialize checksum value to zero.
+* Fetch operation will return random values in checksum field.
+* Test will compare the update and fetch output for pass/fail
+*/
+void
+csum_fault_injection_multiple_extents_tests(void **state)
+{
+	int rc;
+	int i;
+	int update_extents;
+	int fetch_extents;
+	struct csum_test test;
+	/* Setup Test */
+	struct csum_test_params params;
+	uint32_t csums_count_total, csum_count_per_extent, csum_len;
+	/* extent counts for update and fetch. The extents will span the same
+	 * amount of data, the extents themselves will be of different lengths
+	 */
+	const int END = 0;
+	const int table[][2] = {
+	/*	update, fetch  */
+		{1,	1},
+		{1,	4},
+		{2,	4},
+		{4,	4},
+		{4,	1},
+		{4,	2},
+		{END,	END}
+	};
+
+
+	params.total_records = 1024 * 1024 * 64;
+	params.record_bytes = 1;
+	params.csum_bytes = 8; /* CRC64? */
+	params.csum_chunk_records = 1024 * 16; /* 16K */
+	params.use_rand_csum = true;
+
+	csum_test_setup(&test, state, &params);
+
+
+	for (i = 0; table[i][0] != END; i++) {
+		update_extents = table[i][0];
+		fetch_extents = table[i][1];
+
+		printf("Update Extents: %d, Fetch Extents: %d\n",
+		       update_extents, fetch_extents);
+
+		daos_fail_loc_set(DAOS_CHECKSUM_UPDATE_FAIL | DAOS_FAIL_ALWAYS);
+		daos_fail_num_set(1);
+		rc = update(&test, update_extents, i);
+		if (!SUCCESS(rc))
+			fail_msg("Error updating extent with csum: %s\n",
+				 d_errstr(rc));
+
+		daos_fail_loc_set(DAOS_CHECKSUM_FETCH_FAIL | DAOS_FAIL_ALWAYS);
+		daos_fail_num_set(1);
```

Comment:
what is this fail_num_set(1) for?

Reply:
Based on my testing, it is the number of times fault injection flag is tested. 
eg: If we perform DAOS_FAIL_CHECK once, the fail_number is decremented by 1.  
In multiple extent testing, we are doing the DAOS_FAIL_CHECK once. In the single value testing, we are doing DAOS_FAIL_CHECK at twice.

Reply:
Well, FAIL_ALWAYS will always set the failure injection, until fail_loc_reset(). 
Actually, fail_num_set() is used together with DAOS_FAIL_SOME. Just to be clear, 

fail_loc_set(.....DAOS_FAIL_SOME);
daos_fail_num_set(1)

equals to
fail_loc_set(DAOS_FAIL_ONCE).

Reply:
Oh, I see. Thanks. Let me remove the daos_fail_set and use just DAOS_FAIL_ALWAYS alone and re-run the testing.